### PR TITLE
add new metrics

### LINKED ISSUE #723: Feat: better wandb logging
Request to migrate the metrics in https://github.com/NVIDIA-NeMo/NeMo/issues/12865 (PR https://github.com/NVIDIA-NeMo/NeMo/pull/13754) to Megatron-Bridge.

### PR DESCRIPTION
**Pull request by dimapihtar**
fixes #723

### CODE REVIEW

File: src/megatron/bridge/training/config.py [hunk 1]
```diff
@@ -654,6 +654,15 @@ class LoggerConfig:
...
     """If set, calculate and log throughput per GPU."""
 
+    log_throughput_to_tensorboard: bool = False
+    """Enable throughput logging to tensorboard."""
+
+    log_throughput_to_wandb: bool = False
+    """Enable throughput logging to wandb."""
+
+    throughput_window_size: int = 100
+    """Number of batches to use for a rolling average of throughput."""
```
Comment (ananthsub) [inline src/megatron/bridge/training/config.py:711]: why add separate flags for each logger backend? the convention so far has been if wandb is specified and if enabled to log to tensorboard, also log the metric to wandb

File: src/megatron/bridge/training/utils/train_utils.py [hunk 2]
```diff
@@ -54,6 +56,19 @@
...
+    "allocated_bytes.all.current": "current_allocated_mem",
+    "active_bytes.all.current": "current_active_mem",
+    "inactive_split_bytes.all.current": "current_inactive_mem",
+    "reserved_bytes.all.current": "current_reserved_mem",
+    "allocated_bytes.all.peak": "peak_allocated_mem",
+    "active_bytes.all.peak": "peak_active_mem",
+    "inactive_split_bytes.all.peak": "peak_inactive_mem",
+    "reserved_bytes.all.peak": "peak_reserved_mem",
+    "num_alloc_retries": "alloc_retries",
+}
```
Comment (ananthsub) [inline src/megatron/bridge/training/utils/train_utils.py]: this differs from the key names used in L471-492. this may be more of a challenge with merging with MLM later in case those users are expecting those key names eactly

File: src/megatron/bridge/training/utils/train_utils.py [hunk 3]
```diff
@@ -589,27 +632,254 @@ def training_log(
...
+    norm = 0.0
+    optimizer_metrics = {}
+
+    for model_chunk in model:
+        for name, p in model_chunk.named_parameters():
+            if p.main_grad is not None and p.requires_grad:
+                # Always log grad norm as a default metric if it's not specified
+                if f"l2_norm/grad/{name}" not in optimizer_metrics:
+                    param_grad_norm = torch.linalg.vector_norm(p.main_grad)
+                    optimizer_metrics[f"l2_norm/grad/{name}"] = param_grad_norm
```
Comment (ananthsub) [inline src/megatron/bridge/training/utils/train_utils.py]: why the default vs being opt-in?

Comment (dimapihtar) [reply src/megatron/bridge/training/utils/train_utils.py]: I think we just need to remove that comment. It's not deafult since it's configurable.

File: src/megatron/bridge/training/utils/train_utils.py (hunk 2 above)
Comment (dimapihtar) [reply src/megatron/bridge/training/utils/train_utils.py]: changed key names in respect to mlm

**Review by ananthsub — COMMENTED** [review]
LGTM! please add unit tests for the new functions added

File: examples/recipes/llama/pretrain_llama3_8b.py [hunk 4]
```diff
@@ -174,11 +173,6 @@ def main() -> None:
     logger.debug("Starting pretraining...")
     pretrain(config=cfg, forward_step_func=forward_step)
 
-    # Cleanup process group
-    if torch.distributed.is_initialized():
-        torch.distributed.barrier()
-        torch.distributed.destroy_process_group()
-
```
Comment (ananthsub) [inline examples/recipes/llama/pretrain_llama3_8b.py:181]: bad rebase?

Comment (dimapihtar) [reply examples/recipes/llama/pretrain_llama3_8b.py:181]: oh looks like it :D

**Comment by dimapihtar** [comment]
metrics examples for 12-layer model and gbs=16:
image
image
image
image

File: src/megatron/bridge/training/utils/train_utils.py [hunk 5]
```diff
@@ -310,6 +326,8 @@ def training_log(
     num_zeros_in_grad: Optional[int],
     config: ConfigContainer,
     global_state: GlobalState,
+    history_wct: list,
+    model,
```
Comment (ananthsub) [inline src/megatron/bridge/training/utils/train_utils.py]: 1. the other tests in `test_train_utils.py` need to be updated with the new arguments here
2. please update the docstring for the new args added, along with typehints for `history_wct` and `model`

File: src/megatron/bridge/training/utils/train_utils.py [hunk 6]
```diff
@@ -55,6 +57,20 @@
         from megatron.core.utils import local_multi_tensor_l2_norm as multi_tensor_l2norm
 
 
+MEMORY_KEYS = {
```
Comment (ananthsub) [inline src/megatron/bridge/training/utils/train_utils.py]:
```suggestion
MEMORY_KEYS: dict[str, str] = {
```

File: src/megatron/bridge/training/utils/train_utils.py [hunk 7]
```diff
@@ -590,27 +626,253 @@ def training_log(
...
+    +-----------------------------------------------+-----------------------------------------------------+
+    Args:
+        model (Union[MegatronModule, list[MegatronModule]]): megatron model state.
+    Returns:
+        Dictionary with L2 norms for each layer.
+    """
+    norm = 0.0
+    optimizer_metrics = {}
+
+    for model_chunk in model:
```
Comment (ananthsub) [inline src/megatron/bridge/training/utils/train_utils.py:724]: nit: we only handle list[MegatronModule]. i dont think a plain MegatronModule is passed through

File: tests/unit_tests/training/utils/test_log_metrics.py [hunk 8]
```diff
@@ -0,0 +1,131 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
```
Comment (ananthsub) [inline tests/unit_tests/training/utils/test_log_metrics.py]: please add tests here alongside other logging tests: https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/f5a691dd117e790e1d6a1bb96a95585224bff503/tests/unit_tests/training/utils/test_train_utils.py#L30

File: src/megatron/bridge/training/config.py [hunk 9]
```diff
@@ -748,6 +754,18 @@ class LoggerConfig:
     log_memory_to_tensorboard: bool = False
     """Enable memory logging to tensorboard."""
 
+    memory_keys: dict[str, str] = None
```
Comment (ananthsub) [inline src/megatron/bridge/training/config.py]:
```suggestion
    memory_keys: dict[str, str] | None = None
```

File: tests/unit_tests/training/utils/test_train_utils.py [hunk 10]
```diff
@@ -842,14 +956,90 @@ def test_memory_tensorboard_logging(
...
+        memory_report = report_memory(memory_keys=None)
+        assert len(memory_report) == 10
+
+        memory_keys = {
+            "reserved_bytes.all.current": "mem-reserved-bytes",
+            "reserved_bytes.all.peak": "mem-max-reserved-bytes",
+        }
+        excepted_keys = ["mem-reserved-gigabytes", "mem-max-reserved-gigabytes"]
+        memory_report = report_memory(memory_keys=memory_keys)
+        assert list(memory_report.keys()) == excepted_keys
```
Comment (ananthsub) [inline tests/unit_tests/training/utils/test_train_utils.py]:
```suggestion
        expected_keys = ["mem-reserved-gigabytes", "mem-max-reserved-gigabytes"]
        memory_report = report_memory(memory_keys=memory_keys)
        assert list(memory_report.keys()) == expected_keys
```
